stages/org.osbuild.oscap.remediation

```diff
@@ -128,15 +130,88 @@ def main(tree, options):
 
+    if arf_results is not None:
+        # run with command in chroot so full path ok
+        cmd.extend(["--results-arf", f"{data_dir}/{arf_results}"])
+
+    cmd.extend(["--results", f"{data_dir}/{xccdf_results}"])
+
     cmd.append(datastream)
 
-    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False)
+    print("OpenSCAP (evaluate): ", repr(cmd))
+    # we need to
+    # 1) inform the scanner that the system is offline (OSCAP_PROBE_ROOT="")
+    # the variable will make OVAL probes to be less affected by the absence
+    # of dynamic things like /proc /dev etc and also will define the path prefix
+    # (we keep it empty since our process is already chrooted into the image filesystem)
+    # 2) provide important environment variables in OSCAP_CONTAINER_VARS as
+    # the probe does not have access to /proc
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False,
+                         env=dict(os.environ,
+                                  OSCAP_PROBE_ROOT="",
+                                  OSCAP_CONTAINER_VARS="container=bwrap-osbuild"))
 
     # oscap return values are:
     # 0 → success
-    # 2 → -- no error, but some checks/remediation failed
+    # 2 → no error, but some checks/remediation failed
     if res.returncode not in (0, 2):
-        raise RuntimeError("oscap content evaluation and remediation failed")
+        raise RuntimeError("oscap content evaluation failed")
+
+    # run openscap in chroot on the image tree (generate remediation script)
+    cmd = [
+        "/usr/sbin/chroot", tree,
+        "/usr/bin/oscap",
+        "xccdf", "generate"
+    ]
+
+    if verbose_level is not None:
+        cmd.extend(["--verbose", verbose_level])
+
+    if verbose_log is not None:
+        cmd.extend(["--verbose-log-file", f"{data_dir}/{verbose_log}.generate-fix"])
+
+    cmd.extend(["--profile", profile_id])
+    cmd.extend(["fix", "--fix-type", "bash",
+                "--output", f"{data_dir}/{REMEDIATION_SCRIPT}"])
+
+    cmd.append(f"{data_dir}/{xccdf_results}")
+
+    print("OpenSCAP (generate remediation): ", repr(cmd))
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, check=False)
+
+    if res.returncode not in (0, 2):
+        raise RuntimeError("oscap failed to generate remediation script")
+
+    # run the remediation script
+    cmd = [
+        "/usr/sbin/chroot", tree,
+        "/usr/bin/bash",
+        f"{data_dir}/{REMEDIATION_SCRIPT}"
+    ]
+
+    log = None
+    if verbose_log is not None:
+        log = open(f"{tree}/{data_dir}/{verbose_log}.remediation", "w", encoding="utf8")
+
+    print("OpenSCAP remediation script cmd: ", repr(cmd))
+    res = subprocess.run(cmd, encoding="utf8", stdout=sys.stderr, stderr=log, check=False)
+
+    if log is not None:
+        log.close()
+
+    # pack result files, it could save dozens of megabytes
+    cmd = [
+        "/usr/sbin/chroot", tree,
+        "/usr/bin/xz"
```

Comment:
Please note the dependency of this stage on `xz` and `chroot` (should have been noted already :confused: ) in the stage doc text at the top of the file.

Reply:
:grimacing: the `chroot` one was my bad